pkg/elevate/elevate.go

```diff
@@ -0,0 +1,225 @@
+package elevate
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/openshift/backplane-cli/pkg/utils"
+	logger "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+const ElevateBackplaneReasonClusterRole string = "elevate-backplane-cluster-admin-reason"
+const ImpersonationUser string = "backplane-cluster-admin"
+
+var CreateTemp = os.CreateTemp
+var OsRemove = os.Remove
+var ExecCmd = exec.Command
+var NewClientset = kubernetes.NewForConfig
+var BuildConfigFromFlags = clientcmd.BuildConfigFromFlags
+var ReadKubeConfigRaw = utils.ReadKubeconfigRaw
+var WriteKubeconfigToFile = clientcmd.WriteToFile
+
+type ElevateConfig struct {
+	GetCurrentUsername                      func(e ElevateConfig) (string, error)
+	AddImpersonationReasonRoleToCurrentUser func(e ElevateConfig, config api.Config) error
+	AddElevationReasonToRawKubeconfig       func(config api.Config, elevationReason string) error
+	WriteConfigToTempKubeconfig             func(config api.Config) (string, error)
+}
+
+func GetCurrentUsername(e ElevateConfig) (string, error) {
+	cmd := ExecCmd("oc", "whoami")
+	buf := bytes.NewBuffer([]byte{})
+	cmd.Stdout = buf
+
+	err := cmd.Run()
+	if err != nil {
+		return "", err
+	}
+	username := strings.TrimSpace(buf.String())
+	if username == "" {
+		return "", errors.New("Username cannot be empty")
+	}
+
+	return username, nil
+}
+
+func AddImpersonationReasonRoleToCurrentUser(e ElevateConfig, config api.Config) error {
+	clientCfg, err := BuildConfigFromFlags("", clientcmd.NewDefaultPathOptions().GetDefaultFilename())
+	if err != nil {
+		return err
+	}
+	clientCfg.Impersonate.UserName = ImpersonationUser
+
+	clientSet, err := NewClientset(clientCfg)
+	if err != nil {
+		return err
+	}
+	ctx := context.TODO()
+
+	err = clientSet.RbacV1().ClusterRoles().Delete(ctx, ElevateBackplaneReasonClusterRole, metav1.DeleteOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	err = clientSet.RbacV1().ClusterRoleBindings().Delete(ctx, ElevateBackplaneReasonClusterRole, metav1.DeleteOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	role, err := clientSet.RbacV1().ClusterRoles().Create(
+		ctx,
+		&v1.ClusterRole{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: ElevateBackplaneReasonClusterRole,
+			},
+			Rules: []v1.PolicyRule{
+				{
+					Verbs:     []string{"impersonate"},
+					APIGroups: []string{"authentication.k8s.io"},
+					Resources: []string{"userextras/reason"},
```

Comment:
You can create this role in MCC ? https://github.com/openshift/managed-cluster-config/blob/58e869251204af640f8140640b9db5fc3f44eccb/deploy/backplane/elevated-sre/00-impersonate-cluster-admin.ClusterRole.yml#L5

Reply:
It actually has to be done on the SREP user themselves (not the impersonating user). I've tested this on a staging cluster and it works: https://github.com/openshift/managed-cluster-config/pull/1534

Reply:
@hectorakemp Let's add those permissions in https://github.com/openshift/managed-cluster-config/blob/58e869251204af640f8140640b9db5fc3f44eccb/deploy/backplane/elevated-sre/00-impersonate-cluster-admin.ClusterRole.yml#L5 ? so it can be used by any other groups that need this permission.

Reply:
:heavy_check_mark: